src/core/DY.Core.Trace.Manipulation.fst

```diff
@@ -433,18 +433,16 @@ let new_session_id prin =
   let* tr = get_trace in
   return (compute_new_session_id prin tr)
 
-val get_state_aux: principal -> state_id -> trace -> option bytes
-let rec get_state_aux prin sess_id tr =
-  match tr with
-  | Nil -> None
-  | Snoc tr_init (SetState prin' sess_id' content) -> (
-    if prin = prin' && sess_id = sess_id' then
-      Some content
-    else
-      get_state_aux prin sess_id tr_init
-  )
-  | Snoc tr_init _ ->
-      get_state_aux prin sess_id tr_init
+val get_state_aux: principal -> state_id -> tr:trace -> option bytes
+let get_state_aux prin sess_id tr =
+  let is_state_for prin sess_id e =
+    match e with
+    | SetState prin' sess_id' content -> prin = prin' && sess_id = sess_id'
+    | _ -> false
+  in
+  let? state_ts = trace_search_last tr (is_state_for prin sess_id) in
+  let SetState _ _ content = get_entry_at tr state_ts in
+  Some content
 
 /// Retrieve the state stored by a principal at some state identifier.
 
```

Comment:
`get_state` could be implemented as:
```Fstar
let get_state prin sess_id =
  let is_state_for prin sess_id e =
    match e with
    | SetState prin' sess_id' _ -> prin = prin' && sess_id = sess_id'
    | _ -> false
  in
  let* tr = get_trace in
  let state_ts_opt = trace_search_last tr (is_state_for prin sess_id) in
  guard_tr(Some? state_ts_opt);*? // Due to the *? monad not handling options nicely
  let Some state_ts = state_ts_opt in
  let SetState _ _ content = get_entry_at tr state_ts in
  return (Some content)
```
to remove `get_state_aux`. This does seem cleaner to me, though of course `get_state` itself becomes a bit more involved (and the `guard_tr` is needed here, where it isn't in `get_state_aux`).